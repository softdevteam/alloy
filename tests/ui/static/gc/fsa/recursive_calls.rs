#![feature(gc)]
#![feature(negative_impls)]
#![allow(dead_code)]
#![allow(unused_variables)]
include!{"./auxiliary/types.rs"}

impl<T: Debug> Drop for Wrapper<T> {
    fn drop(&mut self) {
        fsa_unsafe(&self, true);
        fsa_safe(&self, true);
    }
}

#[inline(never)]
fn fsa_unsafe<T: Debug>(x: &Wrapper<T>, recurse: bool) {
    if recurse {
        fsa_unsafe(x, false);
    }
    use_val(&x.0); // should fail
}

#[inline(never)]
fn fsa_safe<T: Debug>(x: &Wrapper<T>, recurse: bool) {
    if recurse {
        fsa_safe(x, false);
    }
    use_val(&x); // should pass
}

fn main() {
    Gc::new(Wrapper(NotFinalizerSafe(1)));
    //~^   ERROR: The drop method for `Wrapper<NotFinalizerSafe>` cannot be safely finalized.
}
