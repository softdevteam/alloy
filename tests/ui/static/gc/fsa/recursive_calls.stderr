error: The drop method for `Wrapper<NotFinalizerSafe>` cannot be safely finalized.
  --> $DIR/recursive_calls.rs:31:13
   |
LL |     use_val(&x.0); // should fail
   |             ----
   |             |
   |             a finalizer cannot safely use this `T`
   |             from a drop method because it does not implement `Send` + `Sync`.
...
LL |     Gc::new(Wrapper(NotFinalizerSafe(1)));
   |     --------^^^^^^^^^^^^^^^^^^^^^^^^^^^^- caused by trying to construct a `Gc<Wrapper<NotFinalizerSafe>>` here.
   |
   = help: `Gc` runs finalizers on a separate thread, so drop methods
           must only use values which are thread-safe.

error: aborting due to 1 previous error

